docs/interfaces.md

```diff
@@ -0,0 +1,568 @@
+// Session Requests to TLCC can be issued to invoke tlcc functions such
+// as getMetadata or verifyMSPId.
+
+// TODO How to integrate the the message authenticator?!
+// should we abstract it as signature here even though we may
+// use cmac once a session is established?
+
+message SessionSetup {
+        // TBD input enclave vk?
+        // TLCC Mrenclave?
+        string channel_id = 3;
+}
+
+message SessionSetupResponse {
+        SessionSetup setup = 1;
+        byte context       = 2;         // context that references the session
+}
+
+message SessionClose {
+        byte context = 1;
+}
+
+message SessionCloseResponse {
+}
+
+// TODO add signature or some othere for of message
+// authentication to session request/response.
+
+message SessionRequest {
+        byte context    = 1;
+        byte tx_context = 2;
+        byte nonce      = 3;
+        oneof request {
+                GetMetadataRequest metadata     = 1; 
+                validateMSPRequest validate_msp = 2;
+        }
+}
+
+message SessionRequeatResponse {
+        SessionRequest request = 1;
+        oneof response {
+                GetMetadataResponse metadata     = 1; 
+                validateMSPResponse validate_msp = 2; 
+        }
+}
+
+```
+
+TLCC keeps a local reference of the enclave id. This reference is volatile. Also a reference to the peer is kept to access the ledger and read blocks.
+
+```go
+type TrustedLedgerCC struct {
+	enclave enclave.Stub
+	peer    *peer.Peer
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+# TLCC_Enclave
+
+## Interface:
+Defined in the TLCC EDL.
+```c++
+// creates a new channel
+public int ecall_join_channel(
+        [in, size=gen_len] uint8_t *genesis, uint32_t gen_len);
+
+// pushes next block to tlcc
+public int ecall_next_block(
+        [user_check] uint8_t *block_bytes, uint32_t block_size);
+```
+
+```c++
+// this secure session functions are defined in a secure channel module
+//included from secure_channel.edl
+
+public int session_setup(
+        [in] setup SessionSetup,
+        [out] SessionSetupResponse);
+
+public int session_close(
+        [in] close SessionClose
+        [out] SessionCloseReponse);
+
+public int session_request(
+        [in] request SessionRequest,
+        [out] SessionRequestResponse;
+```
+
+```c++
+// addtionally we may use the state sealing module to persists TLCC state
+// see below the data sealing module.
+include sealed_storage.edl
+```
+
+Session requests are depatched to the following interfaces:
+```c++
+// TODO refine
+// verify state data
+public uint metadata get_state_metadata(
+        const char *namespace,
+        const char *key);
+
+public int get_multi_state_metadata(
+        [in, string] const char *namespace,
+        [in, string] const char *comp_key,
+        [out] uint8_t *metadata);
+
+// verify msp identities
+public bool validate_msp_id(
+        [in, string] const char *msp_id);
+
+// verify msp identities
+public bool can_endorse(
```

Comment:
right, it indeed needs the `enclave_id` and the `chaincode_id`

Reply:
Again, from the `enclave_id` is should be possible to go back to the `chaincode_id`. However, though it appears redundant, it might be useful in the case the enclave is not specialized for a particular chaincode app.